### PR TITLE
[Cache] Doc context aware adapters, aka hierarchical invalidation

### PR DESCRIPTION
**Pull request by nicolas-grekas**

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
I'd happily accept any suggests for making the explanations more detailed. It's hard to write doc when things seem obvious because you wrote the code...

File: components/cache/cache_invalidation.rst [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+several cached items, keeping them in sync can be difficult.
+
+The Symfony Cache component provides three mechanisms to help solve this problem:
+
+- tags based invalidation for managing data dependencies;
+- namespace based invalidation for context and sub-contexts dependend data;
+- expiration based invalidation for time related dependencies.
+
+.. versionadded:: 3.2
+    Tags and namespace based invalidation were introduced in Symfony 3.2.
```
Comment (wouterj) [inline components/cache/cache_invalidation.rst]: I believe it should be "was" here, as it's refering to "X based invalidation", which is singular.

File: components/cache/cache_invalidation.rst [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+Cache invalidation is the process of removing all cached items related to a
+change in the state of your model. The most basic kind of invalidation is direct
+items deletion. But when the state of a primary resource has spread accross
+several cached items, keeping them in sync can be difficult.
+
+The Symfony Cache component provides three mechanisms to help solve this problem:
+
+- tags based invalidation for managing data dependencies;
+- namespace based invalidation for context and sub-contexts dependend data;
+- expiration based invalidation for time related dependencies.
```
Comment (wouterj) [inline components/cache/cache_invalidation.rst]: please use `*` for list items

Comment (wouterj) [inline components/cache/cache_invalidation.rst]: and please start all items with a capitalized word

File: components/cache/cache_invalidation.rst [hunk 3]
```diff
@@ -0,0 +1,104 @@
+.. index::
+   single: Cache Invalidation
+   single: Cache Tags
+   single: Cache Namespaces
```
Comment (wouterj) [inline components/cache/cache_invalidation.rst]: `single: Cache; Namespaces` (same for the other two)

File: components/cache/cache_invalidation.rst [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+To benefit from the tags based invalidation, your responsiblity is to attach the
+proper tags to each cached items. Each tag is a plain string identifier that you
+can use to trigger the removal of all items that had this tag attached to them.
+
+To attach tags to cached items, you need to use the
+:method:`Symfony\\Component\\Cache\\CacheItem::tag` method that is implemented by
+cache items, as returned by cache adapters::
+
+    $item = $cache->getItem('cache_key');
+    // [...]
```
Comment (wouterj) [inline components/cache/cache_invalidation.rst]: `// ...` (remove brackets)

File: components/cache/cache_invalidation.rst [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    $item->tag('tag_1');           // you can add a single tag
+    $item->tag(['tag_2', 'tag_3']); // or several at once
+    $cache->save($item);
+
+Let's assume ``$cache`` implements
```
Comment (wouterj) [inline components/cache/cache_invalidation.rst]: please avoid the first person in the documentation ("let's")

File: components/cache/cache_invalidation.rst [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+:class:`Symfony\\Component\\Cache\\TagAwareAdapterInterface`, you can then
+invalidate the previously cached item by calling e.g.::
```
Comment (wouterj) [inline components/cache/cache_invalidation.rst]:
```rst
If `$cache` implements :class:`Symfony\\Component\\Cache\\TagAwareAdapterInterface`, you
can invalidate the cached items by calling :method:`Symfony\\Component\\Cache\\TagAwareAdapterInterface::invalidateTags`::
```

Comment (nicolas-grekas) [inline components/cache/cache_invalidation.rst]: thanks @wouterj comments addressed

File: components/cache/cache_invalidation.rst [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+:method:`Symfony\\Component\\Cache\\Adapter\\TagAwareAdapterInterface::invalidateTags`
+method by yourself.
+
+The :class:`Symfony\\Component\\Cache\\Adapter\\TagAwareAdapter` class implements
+instantaneous invalidation. It needs one or two cache adapters: the required one
+is used to store cached items; the second one is used to store tags and their
+invalidation version number (conceptually similar to their latest invalidation
+date). When only one adapter is used, items and tags are all stored in the same
+place. By using two adapters, you can e.g. store some big cached items on the
+filesystem or in the database, and keep tags in a Redis database to sync all your
```
Comment (wouterj) [inline components/cache/cache_invalidation.rst]: please remove the serial comma before "and keep"

File: components/cache/cache_invalidation.rst [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+fronts and have very fast invalidation checks::
+
+    $cache = new TagAwareAdapter(
```
Comment (wouterj) [inline components/cache/cache_invalidation.rst]: missing use statement

File: components/cache/cache_invalidation.rst [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        // Adapter for cached items
+        new FilesystemAdapter(),
+        // Adapter for tags
+        new RedisAdapter('redis://localhost')
+    );
+
+Using Cache Namespaces
+----------------------
+
+By using adapters that implement the
```
Comment (wouterj) [inline components/cache/cache_invalidation.rst]: this section is missing some information, at least I can't understand what it's explaining.

 * What are cache namespaces?
 * What is forking?
 * What are "context-dependent variations of your cached items"?
 * When would I use cache namespaces and forks?

File: components/cache/cache_invalidation.rst [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+.. note::
+
+    Invalidating by tags affects all forks.
+
+Using Cache Expiration
+----------------------
+
+If your data is valid only for a limited period of time, you can specify their
+lifetime or their expiration date with the PSR-6 interface, as explained in the
+Cache Items article.
```
Comment (wouterj) [inline components/cache/cache_invalidation.rst]: please add a reference to the cache items article: ``in the :doc:`/components/cache/cache_items` article.``

File: components/cache/cache_pools.rst [hunk 11]
```diff
@@ -102,6 +102,33 @@ the ``\Redis``, ``\RedisArray``, ``\RedisCluster`` or ``\Predis`` classes::
...
+helper allows creating a connection to a Redis server using a DSN configuration::
+
+    $redisConnection = RedisAdapter::createConnection('redis://localhost');
+
+See the method's docblock for more options.
+
+PDO & Doctrine DBAL Cache Adapter
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This adapter stores the cached items a SQL database accessed through a PDO or a
```
Comment (wouterj) [inline components/cache/cache_pools.rst]: missing versionadded directive

**Comment by wouterj** [comment]
Sorry, was distracted for some time during the review of this PR. I've now finished reviewing this article, thanks for documenting this feature!

status: needs work

**Comment by Koc** [comment]
How to configure tag aware cache using framework config?

**Comment by nicolas-grekas** [comment]
@Koc I don't know yet :) would need a PR on FrameworkBundle.

File: components/cache/cache_pools.rst (hunk 11 above)
Comment (nicolas-grekas) [inline components/cache/cache_pools.rst]: added

File: components/cache/cache_invalidation.rst (hunk 9 above)
Comment (nicolas-grekas) [inline components/cache/cache_invalidation.rst]: Explanations updated. More readable now? Can someone understand the concept?

File: components/cache/cache_invalidation.rst [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+you can invalidate the cached items by calling
+:method:`Symfony\\Component\\Cache\\TagAwareAdapterInterface::invalidateTags`::
+
+    // invalidate all items related to `tag_1`
+    $cache->invalidateTags('tag_2');
+
+    // or invalidate all items related to `tag_1` or `tag_3`
+    $cache->invalidateTags(array('tag_1', 'tag_3'));
+
+Of course when you know the cache key, it's better to call
```
Comment (mickaelandrieu) [inline components/cache/cache_invalidation.rst]: I would reword that this way:

Using Cache tags are very useful when you don't need to know the cache key to delete.

And I will mention ``$cache->deleteItem('cache_key');`` in the previous example (self-explained).

WDYT ?

File: components/cache/cache_invalidation.rst [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+and this is where cache tags become useful.
+
+Tag Aware Adapters
+~~~~~~~~~~~~~~~~~~
+
+To store tags, you need to wrap a cache adapter with the
+:class:`Symfony\\Component\\Cache\\Adapter\\TagAwareAdapter` class or implement
+:class:`Symfony\\Component\\Cache\\Adapter\\TagAwareAdapterInterface` and its only
+:method:`Symfony\\Component\\Cache\\Adapter\\TagAwareAdapterInterface::invalidateTags`
+method by yourself.
```
Comment (mickaelandrieu) [inline components/cache/cache_invalidation.rst]: I'm not sure about "by yourself" ^^, the sentence is Ok without it.

File: components/cache/cache_invalidation.rst [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+The :class:`Symfony\\Component\\Cache\\Adapter\\TagAwareAdapter` class implements
+instantaneous invalidation. It needs one or two cache adapters: the required one
```
Comment (mickaelandrieu) [inline components/cache/cache_invalidation.rst]: I had 1 question when I first rode the first sentence : "1 or 2 ?" it's a little disturbing

> You need to use an Adapter to defined where items and tags are stored.
Using a second Adapter allow you to define a specific place to store you tags. For instance, you can store some big cached items on the [...]

WDYT

File: components/cache/cache_invalidation.rst [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+unable to collide with the same key in its parent adapter.
+
+You can use forks everywhere you would otherwise manually prefix your cache keys
+in your code. A typical example is storing different HTML pages per
+``Vary: User-Agent`` HTTP headers: just use one fork per ``Vary`` header.
+
+An added benefit to forks is that clearing them takes nesting into account;
+clearing a parent adapter recursively clears all its forks::
+
+    $childCache->clear();
```
Comment (mickaelandrieu) [inline components/cache/cache_invalidation.rst]: how I can clear a fork cache without "dispatch" the clear to all children ? Is it possible ?

**Comment by mickaelandrieu** [comment]
The docs is clear to me (non-native english).

Big thanks to you @nicolas-grekas !  I want to play with this component now ^^

File: components/cache/cache_invalidation.rst (hunk 15 above)
Comment (nicolas-grekas) [inline components/cache/cache_invalidation.rst]: You can't. Forks are organized in subtrees and clear one branch of the tree clears all sub-trees.
You stil have tags to deal with other schemes of invalidations.

File: components/cache/cache_invalidation.rst [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+
+Cache invalidation is the process of removing all cached items related to a
+change in the state of your model. The most basic kind of invalidation is direct
+items deletion. But when the state of a primary resource has spread accross
+several cached items, keeping them in sync can be difficult.
+
+The Symfony Cache component provides three mechanisms to help solve this problem:
+
+* Tags based invalidation for managing data dependencies;
+* Namespace based invalidation for context dependend data;
```
Comment (tgalopin) [inline components/cache/cache_invalidation.rst]: dependent

**Comment by nicolas-grekas** [comment]
Comments addressed, examples updated. More comments welcomed :)

File: components/cache/cache_invalidation.rst [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+with their parent (esp. the data store connection) but add a ``$namespace``
+prefix to all their keys, in a way that makes a cache key in a forked adapter
+unable to collide with the same key in its parent adapter.
+
+You can use forks everywhere you would otherwise manually prefix your cache keys
+in your code. A typical example is storing different HTML pages per e.g.
+``Vary: User-Agent`` or ``Vary: Accept-Encoding`` HTTP headers::
+
+    $perUserAgentCache = $httpCache->fork($varyingUserAgent);
+    $perAcceptEndoginPerUserAgentCache = $childCache->fork($varyingAcceptEncoding);
```
Comment (tgalopin) [inline components/cache/cache_invalidation.rst]: Encoding

File: components/cache/cache_invalidation.rst [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+
+    // Here, several $pageItem objects can be stored per $someUrl:
+    // one per $varyingUserAgent and per $varyingAcceptEncoding
+    $pageItem = $perAcceptEndoginPerUserAgentCache->getItem($someUrl);
```
Comment (tgalopin) [inline components/cache/cache_invalidation.rst]: Encoding

**Comment by kbond** [comment]
Maybe it is just me but I am not quite understanding the "Namespaces"/"Forkable" feature. Perhaps a more extensive example?

**Comment by nicolas-grekas** [comment]
@kbond see https://www.drupal.org/developing/api/8/cache/contexts for another use case of this feat. Maybe one should first read this page then help me find the right words to explain the feat. ? :)
This is hierarchical invalidation done from the *outside*, (the *inside* being prefixing keys explicitly e.g. `->getItem('foo/bar')`)

File: components/cache/cache_invalidation.rst (hunk 9 above)
Comment (sstok) [inline components/cache/cache_invalidation.rst]: Maybe add an pseudo example how the storage would like like? As a tree.

File: components/cache/cache_invalidation.rst [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+several cached items, keeping them in sync can be difficult.
+
+The Symfony Cache component provides three mechanisms to help solve this problem:
+
+* Tags based invalidation for managing data dependencies;
+* Namespace based invalidation for context dependent data;
+* Expiration based invalidation for time related dependencies.
+
+.. versionadded:: 3.2
+    Tags and namespace based invalidation was introduced in Symfony 3.2.
```
Comment (fabpot) [inline components/cache/cache_invalidation.rst]: were introduced

**Comment by nicolas-grekas** [comment]
I just rewrote entirely this paragraph, any better?

**Comment by Koc** [comment]
Much better now for me

**Comment by kbond** [comment]
I'm understanding it much better as well. The example is great!

**Comment by nicolas-grekas** [comment]
Closing, see https://github.com/symfony/symfony/pull/19521#issuecomment-249273430
